src/gt4py/storage/storage.py

```diff
@@ -39,659 +44,372 @@ def _error_on_invalid_backend(backend):
+            strides, the information is not stored.
+
+    Returns
+    -------
+        NumPy or CuPy ndarray
+            Initialized to 0.0, padded and aligned to provide optimal performance for the given `backend` and
+            `aligned_index`
+
+    Raises
+    -------
+        TypeError
+            If arguments of an unexpected type are specified.
+        ValueError
+            If illegal or inconsistent arguments are specified.
+    """
     storage = empty(
         shape=shape,
         dtype=dtype,
         backend=backend,
-        default_origin=default_origin,
-        mask=mask,
-        managed_memory=managed_memory,
+        aligned_index=aligned_index,
+        dimensions=dimensions,
     )
     storage[...] = 0
     return storage
 
 
 def from_array(
-    data, backend, default_origin, shape=None, dtype=None, mask=None, *, managed_memory=False
-):
+    data: ArrayLike,
+    dtype: DTypeLike = np.float64,
+    *,
+    backend: str,
+    aligned_index: Sequence[int],
+    dimensions: Optional[Sequence[str]] = None,
+) -> Union[np.ndarray, "cp.ndarray"]:
+    """Allocate an array with values initialized to those of `data` with performance-optimal strides and alignment. This
+    copies the values from `data` to the resulting buffer.
+
+
+    Parameters
+    ----------
+        data : `ArrayLike`
+            Source data to be copied, from which also the `shape` of the result is derived.
+        dtype :  DTypeLike, optional
+            The dtype of the resulting `ndarray`
+
+    Keyword Arguments
+    -----------------
+        backend : `str`
+            The target backend for which the allocation is optimized.
+        aligned_index: `Sequence` of `int`
+            Indicate the index of the resulting array that most commonly corresponds to the origin of the compute
+            domain.
+        dimensions: `Sequence` of `str`, optional
+            Indicate the semantic meaning of the dimensions in the provided array. Only used for determining optimal
+            strides, the information is not stored.
+
+    Returns
+    -------
+        NumPy or CuPy ndarray
+            Copy of `data`, padded and aligned to provide optimal performance for the given `backend` and
+            `aligned_index`
+
+    Raises
+    -------
+        TypeError
+            If arguments of an unexpected type are specified.
+        ValueError
+            If illegal or inconsistent arguments are specified.
+    """
+    import gt4py.storage.utils  # prevent circular import
+
     is_cupy_array = cp is not None and isinstance(data, cp.ndarray)
-    xp = cp if is_cupy_array else np
-    if shape is None:
-        shape = xp.asarray(data).shape
+    asarray = gt4py.storage.utils.as_cupy if is_cupy_array else gt4py.storage.utils.as_numpy
+    shape = asarray(data).shape
     if dtype is None:
-        dtype = xp.asarray(data).dtype
+        dtype = asarray(data).dtype
+    dtype = np.dtype(dtype)
```

Comment:
If `cupy.ndarray.dtype` returns `np.dtype` instances, then I guess this is redundant... although most likely a bit safer
```suggestion
```

Reply:
dtype is only derived from `data` if it is `None` in the explicit argument. Otherwise, it can be e.g. `str` instead and needs to be marshalled through `np.dtype`

Reply:
Sure, you're fully right here. My mistake.